pkg/eestream/aesgcm.go

```diff
@@ -0,0 +1,135 @@
+// startingNonce. The monotonically-increasing nonce (that rolls over) is to
+// protect against data reordering.
+//
+// When in doubt, generate a new key from crypto/rand and a startingNonce
+// from crypto/rand as often as possible.
+func NewAesGcmEncrypter(key *[32]byte, startingNonce *[12]byte,
+	encryptedBlockSize int) (Transformer, error) {
+	block, err := aes.NewCipher((*key)[:])
+	if err != nil {
+		return nil, err 
+	}
+	aesgcmEncrypt, err := cipher.NewGCM(block)
+	if err != nil {
+		return nil, err 
+	}
+	if encryptedBlockSize <= aesgcmEncrypt.Overhead() {
+		return nil, err 
+	}
+	return &aesgcmEncrypter{
+		blockSize:     encryptedBlockSize - aesgcmEncrypt.Overhead(),
+		key:           *key,
+		startingNonce: *startingNonce,
+		overhead:      aesgcmEncrypt.Overhead(),
+		aesgcm:        aesgcmEncrypt,
+	}, nil
+}
+
+func (s *aesgcmEncrypter) InBlockSize() int {
+	return s.blockSize
+}
+
+func (s *aesgcmEncrypter) OutBlockSize() int {
+	return s.blockSize + s.overhead
+}
+
+func calcGcmNonce(startingNonce *[12]byte, blockNum int64) (rv [12]byte,
+	err error) {
+	if copy(rv[:], (*startingNonce)[:]) != len(rv) {
+		return rv, Error.New("didn't copy memory?!")
+	}
+	_, err = incrementBytes(rv[:], blockNum)
+	return rv, err
+}
+
+func (s *aesgcmEncrypter) Transform(out, in []byte, blockNum int64) (
+	[]byte, error) {
+	n, err := calcGcmNonce(&s.startingNonce, blockNum)
+	if err != nil {
+		return nil, err
+	}
+
+	ciphertext := s.aesgcm.Seal(out, n[:], in, nil)
+	return ciphertext, nil
+}
+
+type aesgcmDecrypter struct {
+	blockSize     int
+	key           [32]byte
+	startingNonce [12]byte
+	overhead      int
+	aesgcm        cipher.AEAD
+}
+
+// NewAesGcmDecrypter returns a Transformer that decrypts the data passing
+// through with key. See the comments for NewSecretboxEncrypter about
+// startingNonce.
+func NewAesGcmDecrypter(key *[32]byte, startingNonce *[12]byte,
+	encryptedBlockSize int) (Transformer, error) {
+	block, err := aes.NewCipher((*key)[:])
+	if err != nil {
+		return nil, err 
+	}
+	aesgcmDecrypt, err := cipher.NewGCM(block)
+	if err != nil {
+		return nil, err 
+	}
+	if encryptedBlockSize <= aesgcmDecrypt.Overhead() {
+		return nil, Error.New("block size too small")
+	}
+	return &aesgcmDecrypter{
+		blockSize:     encryptedBlockSize - aesgcmDecrypt.Overhead(),
+		key:           *key,
+		startingNonce: *startingNonce,
+		overhead:      aesgcmDecrypt.Overhead(),
+		aesgcm:        aesgcmDecrypt,
+	}, nil
+}
+func (s *aesgcmDecrypter) InBlockSize() int {
+	return s.blockSize + s.overhead
+}
+
+func (s *aesgcmDecrypter) OutBlockSize() int {
+	return s.blockSize
+}
+
+func (s *aesgcmDecrypter) Transform(out, in []byte, blockNum int64) (
+	[]byte, error) {
+	n, err := calcGcmNonce(&s.startingNonce, blockNum)
+	if err != nil {
+		return nil, err
+	}
+
+	plaintext, err := s.aesgcm.Open(out, n[:], in, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return plaintext, nil
```

Comment:
lines 129 through 134 here could be shorted to just

```
return s.aesgcm.Open(...)
```

instead of 

```
pt, err := s.aesgcm.Open(...)
if err != nil { 
  return nil, err
}
return pt, nil
```